uint8 nums
float32[] distance